### PR TITLE
Cleanup cmake files

### PR DESCRIPTION
**Pull request by makxenov**
I would like to consider this PR as a place for discussions first. I feel like our cmake files in crypto3/blueprint/zkllvm require rework. Here are some points for it:
- Files contain a lot of unused code. Huge amount of it was removed from crypto3 by [crypto3#242](https://github.com/NilFoundation/crypto3/pull/242) and [crypto3#233](https://github.com/NilFoundation/crypto3/pull/233). But there are still unused or weird parts, like `cmake_policy`.
- Seems like a lot of projects were initialized as a copy-paste cmake configuration from some existing one. That's the reason why we can see repetitive errors across the projects, and a bunch of unused files like `TargetArchitecture.cmake`, `CheckAVX.cmake` and so on.
- Boost CMake modules used in each project are overcomplicated and non-intuitive. Their macros set some variables that later used outside the macro scope, which is hard to read and debug. Some of them are just incorrect, like `cm_find_package`, that is full of workarounds.
- Statements like `add_library(${CMAKE_WORKSPACE_NAME}_${CURRENT_PROJECT_NAME} INTERFACE)` blur eyes, this customizability is excessive. `add_library(crypto3::blueprint INTERFACE)` is much better.

I think that significant part of current cmake configuration of blueprint could be removed. Also it makes sense to replace usage of Boost cmake modules (expect probably CMTest) with standard cmake library functions. The modules were intended for Boost, they are too complex for us and have bugs.  And standard functions have better documentation and tons of usecases/troubleshooting in public access.

Below I added comments on some removed/replaced parts. I invite you all to discuss this points here. Hope we will get a common understanding of how cmake files should look like and then propagate it to other repos.

### CODE REVIEW

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:7]: All these policies are `NEW` by default since cmake 3.13.  And I don't see any sense in setting them rather then bumping a version.

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:11]: None of the modules from these paths were included

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:50]: Since we are abandoning using blueprint as a submodule, these options can affect only on blueprint tests. So moved them to test subdirectory. If we need some of them activated for targets that are using blueprint, we need to add them to the package config.

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:69]: Procps is unused

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:71]: This list is unused, removing like in https://github.com/NilFoundation/crypto3/pull/242

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:18]: The target will be exported with namespace `crypto3::`, so adding alias target for consistency.

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:136]: Removed heavy and unadaptable `cm_deploy` in favor of manual package config

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:46]: This is probably the most controversial part of the changes. I added a bunch of cmake library functions (`configure_package_config_file`, `write_basic_config_version_file`, `install(FILES ...)`, `install(EXPORT...)`) plus `Config.cmake.in` file instead of single `cm_deploy` call. I still think this is a better solution. Here are some arguments:
- Much easier to debug. All arguments of cmake functions are in front of eyes and defined in the very same file. Comparing to case when Boost modules set some target properties in one function and then extract them in another one to pass inside `install`, so you need to examine whole cmake trace to understand what is going on, the proposed solution is very simple to debug.
- It is more flexible. If we would need to export some option of blueprint, like `BLUEPRINT_PLACEHOLDER_PROOF_GEN`, there is no way to do it with `cm_deploy` without rewriting a bunch of cmake modules. That is because `cm_deploy` generates config file from a unified template. But here we can simply add it to `Config.cmake.in` file.

I would agree that `cm_deploy` is the best option when we had 20+ repositories with complex graph of dependencies. But since we reduced the number to a couple, manual config generation seems preferable. If we want to reduce an amount of code to a minimum, I think we could create some simplified analogue of `cm_deploy` function that would include the commands below.

File: CMakeLists.txt
Comment (Iluvmagick) [inline CMakeLists.txt:28]: Please remove crypto3::all as this breaks my non-flake development environment; and the flake redirects do not work due to cmake errrors.
This makes doing changes to both crypto3 and blueprint at the same time very difficult.

**Review by Iluvmagick — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (Iluvmagick) [reply CMakeLists.txt:28]: Maybe this is possible to fix some other way?
I am unsure if anyone else is testing the 'intended' flake development way; but it does not seem to function for me.

Comment (makxenov) [reply CMakeLists.txt:28]: Seems like we agreed to use nix flakes as a common way of managing dependencies, you can see it already in master. I know it could be painful to switch to it :( 
You can try to run cmake inside `nix develop --override-input nil_crypto3 /path/to/crypto3 .?submodules=1#` shell. It should work now if your local crypto3 is on the current master.

Comment (Iluvmagick) [reply CMakeLists.txt:28]: Now the cmake no longer fails, but it still doesn't work, the local changes are not actually propagated.
e.g. adding
```
should break the build
```
line (with no quotes) to one of my c++ includes inside crypto3 should break the build in blueprint.
But it doesn't; changes are not propagated. I have to run a new `nix develop` command every time I make a change to my other repository.
While developing zkevm I have to juggle both crypto3, blueprint, the actual zkevm repository and the assigner repositories, re-running `nix develop` every time something changes somewhere for each repository.
This is not fun.

File: CMakeLists.txt
Comment (makxenov) [inline CMakeLists.txt:8]: Move inside project command

File: CMakeLists.txt
Comment (makxenov) [reply CMakeLists.txt:28]: @Iluvmagick actually it is still possible to build any project without Nix. But you have to install all the dependencies and rebuild them when needed. Here is an example of a script that may be helpful:
```bash
#!/bin/bash -ex

# CRYPTO3_SRC=/path/to/crypto3
# CM_SRC=/path/to/cmake/modules
# CMAKE_PREFIX_PATH=/desired/install/path/

if [[ ! -d $CRYPTO3_SRC ]]; then
    echo "CRYPTO3_SRC = $CRYPTO_SRC is not a directory"
    exit 1
fi

if [[ ! -d $CM_SRC ]]; then
    echo "CM_SRC = $CM_SRC is not a directory"
    exit 1
fi

if [[ ! -d $CM_SRC ]]; then
    echo "CM_SRC = $CM_SRC is not a directory"
    exit 1
fi

if [[ -z $CMAKE_PREFIX_PATH ]]; then
    echo "CMAKE_PREFIX_PATH is not set"
    exit 1
fi

build_module () {
    MODULE_SRC=$1
    INTERNAL_PREFIX=$2
    cmake $MODULE_SRC -B $MODULE_SRC/build -DCMAKE_INSTALL_PREFIX=$CMAKE_PREFIX_PATH/$INTERNAL_PREFIX
    make -C $MODULE_SRC/build install
}

mkdir -p $CMAKE_PREFIX_PATH
build_module $CM_SRC ""
build_module $CRYPTO3_SRC crypto3
```
Usage:
```
export CMAKE_PREFIX_PATH=/some/desired/path/for/library/installation
CRYPTO3_SRC=/path/to/crypto3 CM_SRC=/path/to/cmake/modules script.sh
cd blueprint
cmake -B build ... # blueprint options
```
But of course if you change something in crypto3, you need to rerun `make -C crypto3/build install` to update installed headers.

Comment (Iluvmagick) [reply CMakeLists.txt:28]: This solution seems to be the same as rerunning `nix develop` every time crypto3 changes, which is what I am already doing. I am not averse to using nix — the tests I've build for blueprint did run on a flake inside `nix develop` — I am averse to having to run commands every time I change something.

[This](https://github.com/NilFoundation/research-excalibur-visualizer/tree/78c1f92d28129d3361fbba4cec2382f589dc74f0) (commit with the version of crypto3 where changes to cmake files haven't broken the process yet) used to work for me.
I can just change the file inside crypto3, and next time I am building a test the changes are reflected in the build with zero additional work required.

Comment (Iluvmagick) [reply CMakeLists.txt:28]: IIRC changing the `crypto3::all` back to the list of modules was basically all I had to do to make the environment functional again, besides adding some boost modules (but that could be done on the level above).
But need to check again, maybe the new cmake changes did make that part unnecessary?

**Review by Iluvmagick — APPROVED** [review]

File: README.md
Comment (akokoshn) [inline README.md:22]: I guess make sense describe `nix develop` command (otherwise users may try build without configure environment)
And how to run test with nix

Comment (makxenov) [reply README.md:22]: Added description for `nix develop` and `nix flake check`. But there is nothing wrong with building without nix environment, all the previous steps do not require nix.

**Review by akokoshn — APPROVED** [review]
What about CI fails?, expected?

**Comment by makxenov** [comment]
> What about CI fails?, expected?

`blueprint_zkevm_zkevm_word_test` was failed, introduced in [this commit](https://github.com/NilFoundation/zkllvm-blueprint/commit/1f43078488b992a52d557fcd393b341dff6a3398#diff-3e2697594c292b72199fe2fdb253e3001554c7caf641141b3fe57350ca2a824b). @Iluvmagick @ETatuzova please fix and please do not merge anything without CI passed.

**Comment by Iluvmagick** [comment]
> > What about CI fails?, expected?
> 
> `blueprint_zkevm_zkevm_word_test` was failed, introduced in [this commit](https://github.com/NilFoundation/zkllvm-blueprint/commit/1f43078488b992a52d557fcd393b341dff6a3398#diff-3e2697594c292b72199fe2fdb253e3001554c7caf641141b3fe57350ca2a824b). @Iluvmagick @ETatuzova please fix and please do not merge anything without CI passed.

This test is broken due to goldilocks field not working as expected; a known issue in multiprecision.
I've decided to not hide the test, although we can comment it out.

**Review by martun — APPROVED** [review]

**Review by ETatuzova — APPROVED** [review]

**Comment by makxenov** [comment]
@Iluvmagick I've disabled the failing test, now we can see one skipped test in the test summary.

File: CMakeLists.txt
Comment (AndreyMlashkin) [inline CMakeLists.txt:63]: Why you're deleting this? We have the same lines in crypto3, we would need to delete it there as well, if there are strong arguments for it
https://github.com/NilFoundation/crypto3/blob/8853748ceffacde6becc9d1ed3f6adacd0f05d29/CMakeLists.txt#L39

File: README.md
Comment (AndreyMlashkin) [inline README.md:26]: we need a section for crypto3 override

File: flake.nix
Comment (AndreyMlashkin) [inline flake.nix:9]: good thing, that we don't need submodules anymore, as there are none left in crypto3

File: CMakeLists.txt
Comment (aleasims) [inline CMakeLists.txt:7]: Should this really be `crypto3_blueprint`? May be we should move away from this `crypto3` prefix, which makes no sense now?

File: CMakeLists.txt
Comment (aleasims) [reply CMakeLists.txt:17]: Same question - isn't it the time to move away from `crypto3` prefix in blueprint?

File: CMakeLists.txt
Comment (aleasims) [inline CMakeLists.txt:71]: Same

File: README.md
Comment (aleasims) [inline README.md:26]: Should be `nix build -L .?submodules=1#`

File: README.md
Comment (aleasims) [inline README.md:32]: Should be `nix develop .?submodules=1#`

File: README.md
Comment (aleasims) [reply README.md:26]: Also in general `-L` is probably not the default flag you want when you simply building library

File: README.md
Comment (aleasims) [inline README.md:38]: Should be `nix flake -L check .?submodules=1#`

File: CMakeLists.txt
Comment (makxenov) [reply CMakeLists.txt:63]: This `find_package` overwrites `Boost_LIBRARIES` variable that is used below in this file. So in case of enabled tests `blueprint` target would be linked with the wrong set of Boost components. I moved this line into `tests` subdirectory. We need to check `crypto3` as well.

File: README.md
Comment (makxenov) [reply README.md]: Added at the end of the file.

File: flake.nix
Comment (makxenov) [reply flake.nix]: Thanks, fixed

File: CMakeLists.txt
Comment (makxenov) [reply CMakeLists.txt:7]: I would like to do it in separate PR and got special approvals on this change. Current changes are just refactoring, they do not break existing naming and compatibility with other repos.

File: README.md
Comment (makxenov) [reply README.md]: Thanks, fixed

File: README.md
Comment (makxenov) [reply README.md:32]: This argument does not affect dev shell since there is no source code copying.

File: README.md
Comment (makxenov) [reply README.md]: Fixed

**Review by aleasims — APPROVED** [review]

**Review by AndreyMlashkin — APPROVED** [review]

**Review by x-mass — APPROVED** [review]

**Pull request merged**
